### ISSUE TITLE
Redeployment of contract without any change of input to constructor / contract code

### ISSUE THREAD

**Issue by playsnook** [wighawag/hardhat-deploy#221]
My deployment script of non-upgradable contracts is as follows:

```js
require('dotenv').config();
const { delayBetweenDeployScripts } = require('../scripts/lib');
const addresses = require('../addresses.json');

const SkillTokenName = process.env.SKILL_TOKEN_NAME;
const SkillTokenSymbol = process.env.SKILL_TOKEN_SYMBOL;

module.exports = async ({
  getNamedAccounts,
  deployments,
  network
}) => {
  const {deploy} = deployments;
  const {deployer} = await getNamedAccounts();
  await deploy('SkillToken', {
    from: deployer,
    gasLimit: 4_000_000,
    gasPrice: 30_000_000_000,
    args: [
      SkillTokenName,
      SkillTokenSymbol,
      process.env.INITIAL_SKILL_SUPPLY_IN_ETHERS,
      18,
      addresses[network.name].ChildChainManagerProxy
    ],
    log: true,
  });
  await delayBetweenDeployScripts();
};
module.exports.tags = ['L2', 'mumbai', 'polygon'];
```

I tried to run `hardhat --network polygon deploy --tags polygon` and redeployment of the contract was initiated, see on the log `deployed at`:

```
Nothing to compile
reusing "DefaultProxyAdmin" at 0xb1c2f36F1aBe8457855a7934E64Bf5C85FFDF6f5
reusing "SnookToken_Implementation" at 0xD7b2CFC2e811cfE5eDb5BFf0289F27c6A92D0A07
deploying "SkillToken" (tx: 0xd0878585a53a9dbff3452dd16c20626cf281931227e66e6e7fe91bf49cce44cc)...: deployed at 0xB09477e16fbE846aD6909a347F4D8DE9DfD86D85 with 2241481 gas
deploying "UniswapUSDCSkill" (tx: 0x26426b942817c9b9a2862374a84a6cf3eb3b18c9ae5ab867ae3f7e3a9c3dd35c)..
```

It has never happened before, the output was like:

```
Compiling 62 files with 0.8.0
Compilation finished successfully
reusing "DefaultProxyAdmin" at 0xb1c2f36F1aBe8457855a7934E64Bf5C85FFDF6f5
deploying "SnookToken_Implementation" (tx: 0xf3341781283c17ea0266d94f06be441f910db60467075633241ba7cbf29f3ced)...: deployed at 0xD7b2CFC2e811cfE5eDb5BFf0289F27c6A92D0A07 with 2291900 gas
---------------------------------------------------------------------------------------
Please execute the following:
---------------------------------------------------------------------------------------

from: 0x11fe914689525618c322fc9cd76333BA15a932a6
to: 0xb1c2f36F1aBe8457855a7934E64Bf5C85FFDF6f5 (DefaultProxyAdmin)
method: upgrade
args:
  - 0x4372597f1c600D86598675DCB6cF5713bB7525Cf
  - 0xD7b2CFC2e811cfE5eDb5BFf0289F27c6A92D0A07

(raw data: 0x99a88ec40000000000000000000000004372597f1c600d86598675dcb6cf5713bb7525cf000000000000000000000000d7b2cfc2e811cfe5edb5bff0289f27c6a92d0a07 )

---------------------------------------------------------------------------------------
reusing "SkillToken" at 0x689f8e5913C158fFB5Ac5aeb83b3C875F5d20309
reusing "UniswapUSDCSkill" at 0x31f528c6c193b27d77B874D1c28d113103AF04BF
reusing "DefaultProxyAdmin" at 0xb1c2f36F1aBe8457855a7934E64Bf5C85FFDF6f5
reusing "Treasury_Implementation" at 0x5CDC6FD7bB53C1646f290A921dDf060e8b2b2CCE
reusing "DefaultProxyAdmin" at 0xb1c2f36F1aBe8457855a7934E64Bf5C85FFDF6f5
```
  
Comparison of solcInput hashes of the previous and redeployed `deployments/polygon/SkillToken.json` shows they are the same:

![screen](https://user-images.githubusercontent.com/85632593/140986792-763a86a4-7c29-4640-847f-5849503040b5.png)

Can you please assist and explain why the plugin decided to redeploy instead of reuse?

Thanks

**Comment by wighawag** [comment]
I think the answer is on the top line that changes that your image cut. could you copy the full diff

**Comment by playsnook** [comment]
These are all the differences.

The top line, that's actually the different deployment addresses:

![image](https://user-images.githubusercontent.com/85632593/141020685-5091aa1a-2a08-44a5-aeb2-4e89db4b9344.png)

Then the second block of differences:

![image](https://user-images.githubusercontent.com/85632593/141020936-f37d2468-04cc-42fe-87f8-f02bd72efb9c.png)
![image](https://user-images.githubusercontent.com/85632593/141020988-88870ca4-196e-4639-bfd9-c10cf1bf9d91.png)
![image](https://user-images.githubusercontent.com/85632593/141021058-09c6312f-e388-4b15-8fb3-90b0704845a4.png)

They relate to receipt, it cannot be a *cause* for the redeploy.

**Comment by wighawag** [comment]
I was referring to the top line of what could be seen in your first post, not the first line of the file. I can now see that the "top lin" was a data field of a log, should not indeed be the trigger.

But you still did not put the full diff and having in text format is far easier to work with,
My bet is that some constructor argument are different which result in the deployment tx data to be different (which is what hardhat-deploy look at.

**Comment by wighawag** [comment]
hmm, I can see the args are the same, so not sure, but having the full diff as text would still help figuring it out

**Comment by wighawag** [comment]
I checked both contract and their deployment tx data is indeed identical

the code in hardhat-deploy is the following: https://github.com/wighawag/hardhat-deploy/blob/d5988f28c7dd2a7adf89607d74faa523eb32535e/src/helpers.ts#L852-L882

Could it be that matic node return null, because it is not synced up. should probably throw an error if the tx for the old contract is non-existant.

Maybe that is what happened ?

**Comment by playsnook** [comment]
Yes, it was the issue. Our provider returned *null* for the transaction with which the contract was deployed. We moved to another node and all got fixed. 
Thank you alot for your help!

**Comment by playsnook** [comment]
This issue is a great lesson I've learnt. 
A question arises: is there a way for dry run of deploy? I would like to see that a contract is going to be re-used or re-deployed before actual deployment. I thought i could write a llittle script to be run before deployment scripts which can repeat functionality of   `helper.ts::fetchIfDifferent()` function. What can you advice?

**Comment by wighawag** [comment]
Nice to know it is fixed, but I ll definitely fix it in hardhat-deploy too, thank you for your report!


For dry-run the best is to run the deploy on a fork, with my template, you can do it via `yarn fork:deploy ploygon` , see : https://github.com/wighawag/template-ethereum-contracts

**Comment by playsnook** [comment]
Thanks alot again!

**Comment by wighawag** [comment]
Added a fix in 0.9.6 so it errors out if transaction not found
